src/librbd/cache/pwl/ssd/WriteLog.cc

```diff
@@ -686,14 +687,6 @@ bool WriteLog<I>::retire_entries(const unsigned long int frees_per_tx) {
       flushed_sync_gen = this->m_flushed_sync_gen;
     }
 
-    //calculate new first_valid_entry based on last entry to retire
-    auto entry = retiring_entries.back();
-    if (entry->is_write_entry() || entry->is_writesame_entry()) {
-      first_valid_entry = entry->ram_entry.write_data_pos +
-        entry->get_aligned_data_size();
-    } else {
-      first_valid_entry = entry->log_entry_index + MIN_WRITE_ALLOC_SSD_SIZE;
-    }
```

Comment:
This should produce the same value for `first_valid_entry` as it's done in the newly added code in line#669.

By removing this, are you implying that this calculation is wrong?

Reply:
@MahatiC . if control_block contains the following log
log1: write_entry
log2: write_entry
log3: non-write_entry
log1/2/3 log_entry_index is the same .  For this control_block, the next control_block_pos is current control_block_pos  + data_length(log1) + data_length(log2) + MIN_WRITE_ALLOC_SSD_SIZE.
But currently, it only  current control_block_pos + MIN_WIRTE_ALLOC_SSD_SIZE.
Is it ok?

Reply:
`entry->ram_entry.write_data_pos` is updated correctly to reflect current `log_entry_index` AFAIK. Do you see an issue with it?

Reply:
Not for write_entry rather than non-write_entry. It only update write_entry's ram_entry.write_data_pos
see:    https://github.com/ceph/ceph/blob/master/src/librbd/cache/pwl/ssd/WriteLog.cc#L837

Reply:
Then I think you should be updating `write_data_pos` in `write_log_entries` function for non-write entries

Reply:
why? for non-data log, why we should update write_data_pos.  Do you think there is anything wrong with my current reforms?

Reply:
sorry. No `write_data_pos` is not needed to be updated for non-write entries.

Then this PR is not needed IMO. For non write entries, the `first_valid_entry` is updated correctly here[0]

[0]https://github.com/ceph/ceph/blob/master/src/librbd/cache/pwl/ssd/WriteLog.cc#L695

Reply:
No. if control_block contain: write_entry, write_entry; non-write_entry.  By your code, first_valid_entry= non-write_entry->log_entry_index + MIN_ALLOC_SIZE. But in fact, it's not correctly.

Reply:
@MahatiC ping!

Reply:
ah ok, I get your point. This change seems correct. But IMO, this needs to be verified by a bug report. I assume the unit tests pass with this change.

Reply:
@MahatiC  yes i already add a tracker.